package-res/resources/web/pentaho/type/Context.js

```diff
@@ -155,15 +240,63 @@ define([
     },
 
     /**
-     * Gets the instance constructor of a type.
+     * Gets, asynchronously, the **configured instance constructor** instance constructor of a type.
+     *
+     * For more information on the `typeRef` argument,
+     * please see [UTypeReference]{@link pentaho.type.spec.UTypeReference}.
+     *
+     * This method can be used even if a generic type metadata specification references
+     * non-standard types whose modules have not yet been loaded by the AMD module system.
+     *
+     * An error is thrown when:
```

Comment:
To test violations of contract or not, that is the question.

I agree that adding code just to ensure a required argument is indeed passed to a function or constructor will put a performance penalty that might be avoided altogether if we assume the documentation is properly followed. That idea about DEBUG variable seems to be great, specially if is only used in a block like:

```js
if(DEBUG){
  // validate arguments
  if (...) throw ...
  if (...) throw ...
  if (...) throw ...
}
```

Personally, I find it currently hard to predict which cases fall into the category of "no need to test that"